pkg/api/latest/codec.go

```diff
@@ -0,0 +1,129 @@
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package latest
+
+import (
+	"fmt"
+	"io"
+	"reflect"
+
+	"k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/api/unversioned"
+	"k8s.io/kubernetes/pkg/conversion"
+	"k8s.io/kubernetes/pkg/runtime"
+	"k8s.io/kubernetes/pkg/runtime/serializer/json"
+	"k8s.io/kubernetes/pkg/runtime/serializer/recognizer"
+)
+
+var (
+	Codecs                CodecFactory
+	UniversalDeserializer runtime.Decoder
+
+	jsonSerializer = json.NewSerializer(conversion.DefaultMetaFactory, api.Scheme, runtime.ObjectTyperToTyper(api.Scheme))
+	yamlSerializer = json.NewYAMLSerializer(conversion.DefaultMetaFactory, api.Scheme, runtime.ObjectTyperToTyper(api.Scheme))
+)
+
+type serializerType struct {
+	AcceptContentTypes []string
+	ContentType        string
+	FileExtensions     []string
+	Serializer         runtime.Serializer
+}
+
+// allows other codecs to be enabled with compile time flags in their own files
+var serializers = []serializerType{
+	{
+		AcceptContentTypes: []string{"application/json"},
+		ContentType:        "application/json",
+		FileExtensions:     []string{"json"},
+		Serializer:         jsonSerializer,
+	},
+	{
+		AcceptContentTypes: []string{"application/yaml"},
+		ContentType:        "application/yaml",
+		FileExtensions:     []string{"yaml"},
+		Serializer:         yamlSerializer,
+	},
+}
+
+func init() {
+	Codecs = CodecFactory{
+		scheme:      api.Scheme,
+		serializers: serializers,
+	}
+	decoders := make([]runtime.Decoder, 0, len(serializers))
+	for _, d := range serializers {
+		decoders = append(decoders, d.Serializer)
+	}
+	UniversalDeserializer = recognizer.NewDecoder(decoders...)
+}
+
+type CodecFactory struct {
+	scheme      *runtime.Scheme
+	serializers []serializerType
+}
+
+type noopEncoder struct {
+	runtime.Decoder
+}
+
+func (e noopEncoder) EncodeToStream(obj runtime.Object, w io.Writer) error {
+	return fmt.Errorf("encoding is not allowed for this codec: %v", reflect.TypeOf(e.Decoder))
+}
+
+// UniversalDecoder returns a runtime.Decoder capable of decoding all known API objects in all known formats. Used
+// by clients that do not need to encode objects but want to deserialize API objects stored on disk. Only decodes
+// objects in groups registered with the scheme. The GroupVersions passed may be used to select alternate
+// versions of objects to return - by default, runtime.APIVersionInternal is used. If any versions are specified,
+// unrecognized groups will be returned in the version they are encoded as (no conversion).
+func (f CodecFactory) UniversalDecoder(versions ...unversioned.GroupVersion) runtime.Decoder {
+	return f.CodecForVersions(noopEncoder{UniversalDeserializer}, versions...)
+}
+
+// CodecFor creates a codec with the provided serializer. If versions are specified, objects are converted to
```

Comment:
Maybe last wins - I don't know that we have a use case yet that requires
priority order.

On Mon, Nov 30, 2015 at 8:05 AM, David Eads <notifications@github.com>
wrote:

> In pkg/api/latest/codec.go
> <https://github.com/kubernetes/kubernetes/pull/17922#discussion_r46140085>
> :
>
> > +}
> > +
> > +func (e noopEncoder) EncodeToStream(obj runtime.Object, w io.Writer) error {
> > +	return fmt.Errorf("encoding is not allowed for this codec: %v", reflect.TypeOf(e.Decoder))
> > +}
> > +
> > +// UniversalDecoder returns a runtime.Decoder capable of decoding all known API objects in all known formats. Used
> > +// by clients that do not need to encode objects but want to deserialize API objects stored on disk. Only decodes
> > +// objects in groups registered with the scheme. The GroupVersions passed may be used to select alternate
> > +// versions of objects to return - by default, runtime.APIVersionInternal is used. If any versions are specified,
> > +// unrecognized groups will be returned in the version they are encoded as (no conversion).
> > +func (f CodecFactory) UniversalDecoder(versions ...unversioned.GroupVersion) runtime.Decoder {
> > +	return f.CodecForVersions(noopEncoder{UniversalDeserializer}, versions...)
> > +}
> > +
> > +// CodecFor creates a codec with the provided serializer. If versions are specified, objects are converted to
>
> So if I pass two different versions for the same group, its a priority
> order?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/17922/files#r46140085>.
>